File: hackerrank/problem solving/algorithms/2 - implementation/24 - viral advertising/with morfo library/viral_advertising.java
// https://www.hackerrank.com/challenges/strange-advertising/problem?isFullScreen=true
// Java 25

import morfo.io.Reader;

void main() {
    int days = Reader.read(Integer::parseInt);
    IO.println(viralAdvertising(days));
}

// n: days
// T: O(n)
// S: O(1) extra space
int viralAdvertising(int days) {
    int acc = 0;
    for (int shared = 5; days > 0; days--) {
        int liked = shared / 2;
        shared = liked * 3;
        acc += liked;
    }
    return acc;
}

File: hackerrank/problem solving/algorithms/2 - implementation/24 - viral advertising/with morfo library/viral_advertising.js
// https://www.hackerrank.com/challenges/strange-advertising/problem?isFullScreen=true

import { idiv, read } from 'morfo';

const days = read(Number);
console.log(viralAdvertising(days));

// n: days
// T: O(n)
// S: O(1) extra space
function viralAdvertising(days) {
  let acc = 0;
  for (let shared = 5; days > 0; days--) {
    const liked = idiv(shared, 2);
    shared = liked * 3;
    acc += liked;
  }
  return acc;
}
